### ISSUE TITLE
Reduce system clock rate to make timing closure easier

### ISSUE THREAD

**Issue by tvannoy** [fpga-open-speech-tools/de10nano_projects#6]
As we currently implement them, the Simulink models only do computation for one system clock cycle (98.304 MHz) every 96 kHz. Students have been having problems meeting timing closure with some of the more complicated Simulink models. 

Since the Audio Mini isn't intended for super high-performance designs, we probably can get away with a much slower system clock than 98.304 MHz. Using a slower clock should make meeting timing closure somewhat easier.

**Comment by rksnider** [comment]
Good point.

Ross

**Comment by tvannoy** [comment]
In support of this, I just recompiled the sound effects project and had a timing closure failure with a slack of around -2.6. 

Changing this will also require some changes in the Simulink models.

**Comment by rksnider** [comment]
Can we script changing the PLL clock output?   We could then change the
system clock for the data plane to be some lower multiple of the sample
rate and do this automatically from a Matlab initialization script.
However, we would need to feed this multiple (and change it to be a
multiple of the Mclock) to the PLL.

**Comment by tvannoy** [comment]
I believe that should be possible. Dylan had to generate a pll for the Arria 10 from the command line for some reason (it already existed in the system). I see no reason that we couldn't create a tcl script that changes the pll output clock rate. @fe-wickham what do you think?

It would be nice to automatically make the clock rates match between the Simulink model and the Quartus project. Now that we can generate Quartus projects from Simulink, this should be how it's done.

**Comment by fe-wickham** [comment]
The PLL clock output could definitely be scripted. So we would want the PLL to match the Matlab value mp.Fs_system aka "system_frequency_Hz" in the JSON file?

**Comment by rksnider** [comment]
Yes, that would be good.  We could then have a Matlab function determine
the minimum system clock frequency that would work.

**Comment by tvannoy** [comment]
> 
> 
> So we would want the PLL to match the Matlab value mp.Fs_system aka "system_frequency_Hz" in the JSON file?

Yes, that's my thinking. I don't think we need any sort of Matlab function to determine the minimum system clock frequency since we specify that frequency manually in the Matlab init scripts already. Maybe I'm missing something that Ross is thinking, though.

**Comment by rksnider** [comment]
When there are rate conversion blocks in the model, the needed clock rate
could be faster than the sample rate or what the user has specified.   This
could be determined by viewing all the sample rates in the model and
picking the fastest one (need to figure out how to parse all the sample
rates out of the model).

**Comment by rksnider** [comment]
I can do this when I parse the model

https://itectec.com/matlab/matlab-how-to-obtain-the-inherited-sampling-rate-of-a-discrete-simulink-block-from-the-matlab-command-line/

**Comment by tvannoy** [comment]
That makes sense. Our models with rate conversions have never had issues
before because we have always manually ensured that the system clock is
faster than the fastest required rate.

If we are going to always choose the lowest possible system rate based off
the model, then we could probably remove the ability to set the system rate
in the init scripts, as that will just get overwritten anyway.

**Comment by rksnider** [comment]
OK, let's plan on removing this for now to make it simpler for the
developer.  I'll need to create a function something like
set_system_clock_rate.m and then set it before we create the Avalon JSON
file.

**Comment by tvannoy** [comment]
That sounds good to me. Now we're being like Mathworks and choosing the system rate for our users 😉 

I think this should work well. We could change our current "system rate" to be a variable "simulation rate" if we want. We will need to set a simulation rate either way; I'm not sure how running a simulation works if the simulation rate is slower than the fastest sampling rate due to a rate transition.

**Comment by rksnider** [comment]
Ouch :).    At some point we may want a manual override option....

**Comment by tvannoy** [comment]
A flag of whether or not to let us choose the system rate would be good. If a user chooses to manually enter the system rate, we can alert them if the needed system rate, determined by your Simulink parsing, is faster than what they set.
